freeipa/src/main/java/com/sequenceiq/freeipa/service/user/UmsUsersStateProvider.java

```diff
@@ -0,0 +1,130 @@
+package com.sequenceiq.freeipa.service.user;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.auth.ThreadBasedUserCrnProvider;
+import com.sequenceiq.cloudbreak.auth.altus.Crn;
+import com.sequenceiq.cloudbreak.auth.altus.GrpcUmsClient;
+import com.sequenceiq.freeipa.api.v1.freeipa.user.model.Group;
+import com.sequenceiq.freeipa.api.v1.freeipa.user.model.User;
+import com.sequenceiq.freeipa.service.user.model.UsersState;
+import com.sequenceiq.freeipa.util.CrnService;
+
+@Service
+public class UmsUsersStateProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UmsUsersStateProvider.class);
+
+    @Inject
+    private ThreadBasedUserCrnProvider threadBaseUserCrnProvider;
+
+    @Inject
+    private CrnService crnService;
+
+    @Inject
+    private GrpcUmsClient umsClient;
+
+    public UsersState getUsersState(String environmentCrn) {
+        final String actorCrn = threadBaseUserCrnProvider.getUserCrn();
+        final String accountId = crnService.getCurrentAccountId();
+
+        List<com.cloudera.thunderhead.service.usermanagement.UserManagementProto.User> users =
+                umsClient.listUsers(actorCrn, accountId, Optional.empty());
+        LOGGER.info("Found {} users", users.size());
+        List<User> convertedUsers = users.stream()
+                .map(user -> umsUserToUser(actorCrn, user, environmentCrn))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        List<com.cloudera.thunderhead.service.usermanagement.UserManagementProto.MachineUser> machineUsers =
+                umsClient.listMachineUsers(actorCrn, accountId, Optional.empty());
+        LOGGER.info("Found {} machine users", machineUsers.size());
+        List<User> convertedMachineUsers = machineUsers.stream()
+                .map(machineUser -> umsMachineUserToUser(actorCrn, machineUser, environmentCrn))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        Set<User> allUsers = new HashSet<>();
+        allUsers.addAll(convertedUsers);
+        allUsers.addAll(convertedMachineUsers);
+
+        Set<Group> groups = allUsers.stream()
```

Comment:
If I'm reading this right, this set will only include groups to which at least one user belongs. Or, if allUsers is changed to include only users that are authorized to access this environment, the set will only include groups with at least one authorized user in the env.

Is this what we want? This is how it worked in Altus DW, which used Sentry; Sentry allows you to define access policies that reference non-existent groups. But my understanding is that Ranger doesn't allow that.

Reply:
captured in DISTX-186